extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,236 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {expandTemplate} from '../../../src/string';
+import {dev, user} from '../../../src/log';
+import {fromClass} from '../../../src/service';
+import {isArray} from '../../../src/types';
+import {base64Encode} from '../../../src/utils/base64';
+import {isFiniteNumber} from '../../../src/types';
+import {cryptoFor} from '../../../src/crypto';
+
+const TAG = 'Analytics.Variables';
+
+
+/**
+ * @typedef {string|!Array<string>}
+ */
+let VariableFilterIODef;
+
+/**
+ * @typedef {function(...VariableFilterIODef):VariableFilterIODef}
+ */
+let VariableFilterDef;
+
+function substrFilter() {
+  user().assert(typeof arguments[0] == 'string',
+    'substr filter can only accept strings.');
+
+  const start = Number(arguments[1]);
+  const length = Number(arguments[2]);
+  user().assert(isFiniteNumber(start),
+    'Start index ' + start + 'in substr filter should be a finite number');
+  user().assert(isFiniteNumber(length),
+    'Length ' + length + ' in substr filter should be a finite number');
+  return arguments[0].substr(start, length);
+}
+
+function defaultFilter() {
```

Comment:
Any ideas on how to handle check-types warnings like following: 
```
extensions/amp-analytics/0.1/variables.js:69: ERROR - actual parameter 2 of VariableService$$module$extensions$amp_analytics$0_1$variables.prototype.register_ does not match formal parameter
found   : function (string, string, string): Promise<VariableFilterIODef$$module$extensions$amp_analytics$0_1$variables>
required: function (...(Array<string>|string)): (Array<string>|string)
    this.register_('substr', substrFilter);
```
/cc @jridgewell